Add gas parameters for event validation + calibration harnesses.

raulk opened the pull request:
Closes #1109.

---

## Changes

- Sets the gas parameters for event validation starting from nv18. See below for more details on the choice of gas parameters.
- Adds a harness for event gas calibration under `testing/calibration`.
	- It offers two modes: (a) target a specific encoded event size, (b) generate specific event shapes determined by the number of entries, key length, value length.
	- We currently use (b) to calibrate gas parameters by generating events in the same shape as the EVM runtime actor does for LOG{0..4} opcodes.
		- That is, events with 1-5 entries, with 32 byte values, and with 2-char keys (although the `data` entry carries a 1-char key: `d`).
		- Note that the EVM runtime drops leading zeros in values, so our calibration targets the worst-case scenario where all topics and the payload have no leading zeros.
	- There's a case for simply dropping (a), but I've kept it because it may come in handy later stages of the FVM roadmap. It compiles but it never runs.
- Caps the maximum event length to 1024 bytes. This limit can be disarmed by activating the `gas-calibration` feature.

## Choice of gas parameters

These parameters hit the 10gas/ns baseline consistently, with a slight overestimation to cover for the worst case scenario reasonably well (just under 10gas/ns). These parameters do lead to a larger overestimation on the 1-entry case (LOG0 -- pretty rare anyway), but I'm happy living with that.

| <br/>entries | <br/>size | <br/>compute\_gas | count<br/> | mean<br/> | std<br/> | min<br/> | 25%<br/> | 50%<br/> | 75%<br/> | max<br/> |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1entries | 40 | 2846000 | 490.0 | 14.264802 | 2.191890 | 8.131429 | 14.978947 | 14.978947 | 15.811111 | 16.741176 |
| 2entries | 79 | 3914600 | 490.0 | 12.859819 | 1.223028 | 9.103721 | 12.233125 | 13.498621 | 13.498621 | 14.498519 |
| 3entries | 118 | 4983200 | 490.0 | 12.285889 | 1.019456 | 9.384557 | 12.458000 | 12.777436 | 12.777436 | 13.113684 |
| 4entries | 157 | 6051800 | 490.0 | 11.927226 | 0.987889 | 9.019076 | 12.103600 | 12.350612 | 12.607917 | 12.876170 |
| 5entries | 196 | 7120400 | 490.0 | 11.596974 | 0.932503 | 8.466587 | 11.847587 | 12.048054 | 12.048054 | 12.255422 |

Here is a visual depiction of the above as emitted from the calibration tools:

![image](https://user-images.githubusercontent.com/1101242/216631308-56050188-31d6-454f-8090-999eef8eccc0.png)

Here's a closer look at the gas distribution in a violin plot:

![gas_distribution](https://user-images.githubusercontent.com/1101242/216634035-3107e6e3-211a-4040-93ba-c5f7715a2862.png)

## Future direction

This exercise has highlighted the fact that CBOR is not a good choice for an exchange format at the ABI level dealing with (future) untrusted inputs. We found that the costs scale with the number of entries (CBOR list items), which we can't learn prior to parsing the event in some way (we could do an O(n) prevalidation though). It's worth revising this choice going forward. Any changes to the ABI will not have externally visible impacts.

raulk commented on fvm/src/kernel/default.rs:
```diff
@@ -985,18 +985,20 @@ where
+                    Err(syscall_error!(IllegalArgument; "panic when decoding event cbor from actor").into())
+                }
+            };
+            t.stop();
```
Now capturing a reading even on deserialization errors.

Stebalien replied:
I'd be careful about that. We usually don't time _failures_ because it can mess up the number's we're looking for (e.g., we can stop half-way through parsing something).

raulk replied:
I think this makes sense here as long as the test harness is controlling for when it's sending malformed input and when it isn't. This allows us to calibrate the costs for invalid CBOR input, although that will be less relevant if/when we move away from the CBOR-based ABI.

Stebalien replied:
Yeah, fair enough. And the gas charges here are specifically about parsing.

raulk commented on testing/calibration/contract/fil-gas-calibration-actor/src/lib.rs:
```diff
@@ -244,14 +294,22 @@ fn random_mutations(data: &mut Vec<u8>, seed: u64, n: usize) {
     std::iter::repeat_with(move || {
-        seed = a * seed + c;
-        seed
+        seed = Wrapping(a) * seed + Wrapping(c);
```
I'm not entirely sure why, but the lack of wrapping arithmetics here caused weird side effects during ASCII string generation, which led to weird characters. This is more correct.

raulk commented on testing/calibration/src/lib.rs:
```diff
@@ -24,7 +24,7 @@ use num_traits::Zero;
 
 pub const WASM_COMPILED_PATH: &str =
-    "../../target/debug/wbuild/fil_gas_calibration_actor/fil_gas_calibration_actor.compact.wasm";
+    "../../target/release/wbuild/fil_gas_calibration_actor/fil_gas_calibration_actor.compact.wasm";
```
This was the culprit for some major headaches. We run the harness in `--release` profile, which means that we recompile contract changes under the `release` profile. Unfortunately the calibration harness was picking the actor up from the `debug` profile, which usually contains a stale version.

@Stebalien -- this might have or might have not affected your readings.

Stebalien replied:
It shouldn't matter. I was timing FVM operations, not things within the actors.

raulk replied:
Correct, it should not affect the readings. However, in my case this ended up picking stale actor code more than once, leading to wrong conclusions. So wanted to give you a heads-up just in case.

Stebalien replied:
Oh.... Damn, I see. I think rust-analyzer's background "let me recompile everything for you and melt a hole in your lap" helped there.

aakoshh commented on testing/calibration/contract/fil-gas-calibration-actor/src/lib.rs:
```diff
@@ -35,6 +38,8 @@ pub enum Method {
     OnRecoverSecpPublicKey,
     /// Measure sends
     OnSend,
+    /// Emit events variying the number of entries, and with variable lengths for keys and values.
```
```suggestion
    /// Emit events varying the number of entries, and with variable lengths for keys and values.
```

raulk replied:
Thanks!

Stebalien commented on fvm/src/gas/price_list.rs:
```diff
@@ -269,13 +269,25 @@ lazy_static! {
+        // storage costs of such actions.
         event_accept_per_index_element: ScalingCost {
             flat: Zero::zero(),
             scale: Zero::zero(),
```
Or we can just charge up-front in validation, I guess.

Stebalien replied:
But this is important as it factors into the cost to "commit" events.

Stebalien commented on fvm/src/gas/price_list.rs:
```diff
@@ -890,17 +902,28 @@ impl PriceList {
         GasCharge::new(
             "OnActorEventAccept",
-            (memcpy * 3u32) + alloc,
+            memcpy + alloc + hash,
```
To make the benchmarks work out, we should:

1. Put a single memcpy + alloc in the first field.
2. Put the second two copies + alloc + hashing in the second field, as these costs are _deferred_, so the benchmarking won't cover them.

raulk replied:
Done, and updated the benchmarks.

raulk commented:
I've updated the benchmarks to cover power of two byte array lengths in the range 32..8KiB in the last entry in the event, which in EVM land will correspond to the event data (which is of arbitrary length). The longer that value gets, the higher the overestimation. We will retain these parameters because the award us good security, and are not too oversized for reasonable inputs (and we prefer to encourage reasonable inputs!).

![validate](https://user-images.githubusercontent.com/1101242/216728281-cf7be4f7-77a8-44ca-b2e6-8702b5b598c4.png)

Furthermore, @Stebalien and I revised the theoretical accept costs and made some adjustments (increased to 2 memory allocs, and added cost of hashing). These costs are also somewhat overestimated but award good security (this diagram shows only 1 memcpy and 1 alloc; hiding all deferred costs that happen when committing the events).

![accept](https://user-images.githubusercontent.com/1101242/216730233-1958bd45-0477-4ed8-9c43-38b6d5d7cd6a.png)
